### ISSUE TITLE
Apple Silicon

### ISSUE THREAD

**Issue by isuruf** [conda-forge/conda-forge.github.io#1126]
- [x] Ask for support for uploading to anaconda.org
- [x] Build LLVM 11.0.0.rc1  and the stack on osx-64 and upload to `conda-forge/label/llvm_rc` - https://github.com/conda-forge/clang-compiler-activation-feedstock/pull/29
    - [x] https://github.com/conda-forge/llvmdev-feedstock/pull/96
    - [x] https://github.com/conda-forge/clangdev-feedstock/pull/106
    - [x] https://github.com/conda-forge/libcxx-feedstock/pull/66
    - [x] https://github.com/conda-forge/openmp-feedstock/pull/39
    - [x] https://github.com/conda-forge/compiler-rt-feedstock/pull/23
    - [x] https://github.com/conda-forge/lld-feedstock/pull/19
- [x] https://github.com/conda-forge/cctools-and-ld64-feedstock/pull/23
- [x] TAPI TBD-v4. Do one of,
   - [x] Make `cctools` use the system linker from osx-64
   - [x] Wait for Apple to open source new TAPI
   - [ ] Patch `cctools` to use `libLLVMTextAPI`
   - [ ] Use `lld`
- [x] New conda release (https://github.com/conda/conda/pull/10128) - https://github.com/conda-forge/conda-feedstock/pull/123
- [x] New conda-build release (https://github.com/conda/conda-build/pull/4005, https://github.com/conda/conda-build/pull/4004, https://github.com/conda/conda-build/pull/3995, https://github.com/conda/conda-build/pull/3997) - https://github.com/conda-forge/conda-build-feedstock/pull/146
- [x] Update pinnings https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/727
- [x] Update constructor
- [x] Update automake https://github.com/conda-forge/automake-feedstock/pull/26
- [x] Update libtool https://github.com/conda-forge/libtool-feedstock/pull/19
- [x] Update ci-setup https://github.com/conda-forge/conda-forge-ci-setup-feedstock/pull/117
- [x] codesigning
  - [x] update conda-feedstock with https://github.com/conda/conda/pull/10260
  - [x] update conda-standalone-feedstock with https://github.com/conda/conda/pull/10260
  - [x] update cctools-and-ld64-feedstock with https://github.com/tpoechtrager/cctools-port/pull/89
  - [x] patch install_name_tool in cctools-port to run ldid

All of the above are independent tasks.

I could use some help in task 2 if anybody is up for it.

cc @conda-forge/core, @chenghlee

**Comment by beckermr** [comment]
We had another issue open for this @isuruf. Can you please close it or combine it here?

It was called arm fretting or some such thing.

**Comment by beckermr** [comment]
> I could use some help in task 2 if anybody is up for it.

Which one is task 2? The new LLVM release?

**Comment by isuruf** [comment]
Yes

**Comment by jjhelmus** [comment]
> * Ask for support for uploading to anaconda.org

I was able to upload and download `osx-arm64` packages to anaconda.org without problem although I have not gotten to the point where I can do from directly from an Apple Silicon machine yet.  An example: https://anaconda.org/jjhelmus/zlib/files?channel=osx-arm64

**Comment by jakirkham** [comment]
One question I've been wondering about, do we want to treat this as a new architecture or do we want to build packages that support both architectures (`x86` and `arm64`)? It sounded like the latter is also possible. So wanted to check.

**Comment by jjhelmus** [comment]
A quick note on anaconda.org support.  Channels backed by a CDN (e.g. conda-forge) will NOT have osx-arm64 packages in the CDN served repository.  That will require a small change in the logic that syncs the packages with the CDN.

**Comment by isuruf** [comment]
> One question I've been wondering about, do we want to treat this as a new architecture or do we want to build packages that support both architectures (x86 and arm64)? 

Yes definitely as a new architecture.

> It sounded like the latter is also possible.

Not without hacking build systems to use `lipo`.

**Comment by isuruf** [comment]
As to the linker issue, I tried several options and none of them work except using the system linker from xcode 12.
1. lld's old Mach-O backend is not maintained anymore.
2. lld's new Mach-O backend has less features than we need.
3. LLVM's TextAPI library and Apple's TAPI library have incompatible interfaces.

**Comment by beckermr** [comment]
good work!

**Comment by isuruf** [comment]
First C package PR: https://github.com/conda-forge/zlib-feedstock/pull/38
First C++ package PR: https://github.com/conda-forge/libcxx-feedstock/pull/67

@jjhelmus, can you check that the `zlib` package works fine?
@conda-forge/bot, can we start a migration with `python` and its dependencies as a start?

**Comment by CJ-Wright** [comment]
Sure, what does the migration need to encompass? Is this similar to the arch rebuild or more like the pypy rebuild?

**Comment by isuruf** [comment]
It's exactly like the `arch` rebuild, but with a couple of small changes.
1. add the following in `conda-forge.yml`
```
build_platform:
  osx_arm64: osx_64
test_on_native_only: True
```

2. Use a different set of packages from `osx_arm64.txt` (Not `arch_rebuild.txt`)

**Comment by CJ-Wright** [comment]
Let me see what I can do.

**Comment by jakirkham** [comment]
Would that build both the `x86_64` packages and `arm64` packages?

**Comment by CJ-Wright** [comment]
@isuruf is there a help team to contact?

**Comment by isuruf** [comment]
I can create one with me. `conda-forge/help-osx-arm64`. If anybody else would like to join, please let me know. (Or add yourself if you are in core)

**Comment by beckermr** [comment]
@isuruf i'd be best to add it as a subteam of staged-recipes

**Comment by isuruf** [comment]
I'm not sure how to do that. I already created one https://github.com/orgs/conda-forge/teams/help-osx-arm64. Can you fix it or delete it and recreate it?

**Comment by beckermr** [comment]
done!

**Comment by CJ-Wright** [comment]
Is there a list of packages to avoid?

**Comment by CJ-Wright** [comment]
@isuruf in zlib why was `linux_armv71` removed from the `conda-forge.yml`?

**Comment by beckermr** [comment]
> Is there a list of packages to avoid?

We can't build anything that needs fortran for starters.

**Comment by isuruf** [comment]
> Is there a list of packages to avoid?

`build` and `run` requirements should be avoided. Only `host`. `build` is from `osx-64` platform and we don't run tests.

>  in zlib why was linux_armv71 removed from the conda-forge.yml?

Because we've removed the docker image for `linux_armv7l` and the new `conda-forge-ci-setup` files are not updated for that platform.

**Comment by CJ-Wright** [comment]
@beckermr can you give me some context around why [this](https://github.com/regro/cf-scripts/blob/master/conda_forge_tick/make_graph.py#L159) clause is needed?

**Comment by isuruf** [comment]
> We can't build anything that needs fortran for starters.

Yes. I don't know how the migrator would handle that node, but if it doesn't stop when it can't build gfortran, an explicit condition needs to be added.

**Comment by CJ-Wright** [comment]
I'm interested in if I need to duplicate [this](https://github.com/regro/cf-scripts/blob/master/conda_forge_tick/migrators/arch.py#L27)

**Comment by isuruf** [comment]
@CJ-Wright, those are all `build` dependencies and `build` dependencies should not be looked at for `osx-arm64` migration.

**Comment by beckermr** [comment]
> @CJ-Wright, those are all build dependencies and build dependencies should not be looked at for osx-arm64 migration.

Except if you need a fortran compiler in build. Then the package cannot be migrated.

> @beckermr can you give me some context around why this clause is needed?

That is a from a change in smithy. We moved the names of the ci support files on linux from `linux_*` to `linux_64_*`. So for some repos that have not been rerendered, we only infer the 64 if it is not given.

**Comment by isuruf** [comment]
> A quick note on anaconda.org support. Channels backed by a CDN (e.g. conda-forge) will NOT have osx-arm64 packages in the CDN served repository. That will require a small change in the logic that syncs the packages with the CDN.

Thanks @jjhelmus for letting us know. Saw this issue in https://github.com/conda-forge/tk-feedstock/pull/43

**Comment by isuruf** [comment]
Currently blocked by https://github.com/conda/conda-build/issues/3155 at libssh2.

**Comment by jjhelmus** [comment]
I pushed a few of the right buttons and the CDN now includes `osx-arm64` packages.  To verify:
```
wget https://conda-static.anaconda.org/conda-forge/osx-arm64/repodata.json
```

I believe that when the repodata patch package include instructions for patching this subdir (empty instructions are fine) then current_repodata.json will be availabled.

**Comment by beckermr** [comment]
thank you @jjhelmus!

**Comment by jjhelmus** [comment]
conda-forge/conda-forge-repodata-patches-feedstock/pull/86

**Comment by isuruf** [comment]
Could use some help in fixing some issues
1. https://github.com/regro/cf-scripts/issues/1222
2. https://github.com/regro/cf-scripts/issues/1220
3. https://github.com/regro/cf-scripts/issues/1221

cc @erykoff

**Comment by jakirkham** [comment]
If people are looking to try things out and provide feedback, have we reached where they could get started. If so, do we have some instructions for them to follow? How should they share feedback?

**Comment by bollwyvl** [comment]
Thanks for hustling so hard on all this, folks! We're now down to where most folk seem pretty happy. Had a couple folks asking on gitter/elsewhere on _when will package X_ be available. Seems like something on `knowledge_base.rst` would give them someplace to land (and hopefully follow-up with testing on a native box). PR incoming!
